File: Agora-RTM-Tutorial-macOS-Objective-C/Agora-RTM-Tutorial/ChatViewController.m
//
//  ChatViewController.m
//  Agora-Rtm-Tutorial
//
//  Created by CavanSu on 2019/2/19.
//  Copyright © 2019 Agora. All rights reserved.
//

#import "ChatViewController.h"
#import "AgoraRtm.h"
#import "MessageCell.h"
#import "ImageMessageCell.h"

@interface ChatViewController () <AgoraRtmDelegate, AgoraRtmChannelDelegate, NSTextFieldDelegate, NSTableViewDataSource, NSTableViewDelegate>
@property (weak, nonatomic) IBOutlet NSTextField *inputTextField;
@property (weak, nonatomic) IBOutlet NSTableView *tableView;

@property (nonatomic, strong) AgoraRtmChannel *rtmChannel;
@property (nonatomic, strong) NSMutableArray *list;
@end

@implementation ChatViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self ifLoadOfflineMessages];
    [AgoraRtm updateDelegate:self];
}

- (void)viewDidAppear {
    [super viewDidAppear];
    switch (self.mode.type) {
        case ChatTypePeer:  self.title = self.mode.name; break;
        case ChatTypeGroup: self.title = self.mode.name; [self createChannel:self.mode.name]; break;
    }
}

- (void)viewWillDisappear {
    [super viewWillDisappear];
    [self leaveChannel];
}

#pragma mark - Send Message
- (void)sendMessage:(NSString *)message {
    AgoraRtmMessage *rtmMessage = [[AgoraRtmMessage alloc] initWithText:message];
    [self sendRtmMessage:rtmMessage];
}

- (IBAction)sendImage:(id)sender {
    NSString *imagePath = [[NSBundle mainBundle] pathForResource:@"image" ofType:@"png"];
    long long requestId;
    
    __weak ChatViewController *weakSelf = self;
    [AgoraRtm.kit createImageMessageByUploading:imagePath withRequest:&requestId completion:^(long long requestId, AgoraRtmImageMessage *message, AgoraRtmUploadMediaErrorCode errorCode) {
        
        if(errorCode != AgoraRtmUploadMediaErrorOk) {
            NSString *alert = [NSString stringWithFormat:@"send image message error: %ld", (long)errorCode];
            [weakSelf showAlert:alert];
            return;
        }
        
        NSImage *thumbnailImage = [weakSelf generateThumbnail:imagePath toByte:8 * 1024];
        NSData *imageData = [weakSelf getJPGImageData:thumbnailImage];
        if(imageData != nil){
            message.thumbnail = imageData;
            message.thumbnailWidth = thumbnailImage.size.width;
            message.thumbnailHeight = thumbnailImage.size.height;
        }
    
        [weakSelf sendRtmMessage:message];
    }];
}

- (NSData *)getJPGImageData:(NSImage *)image {

    NSData *imageData = [image TIFFRepresentation];
    if (!imageData) {
        return nil;
    }

    NSBitmapImageRep *imageRep = [NSBitmapImageRep imageRepWithData:imageData];
    NSDictionary *imageProps = [NSDictionary dictionaryWithObject:[NSNumber numberWithFloat:1] forKey:NSImageCompressionFactor];
    imageData = [imageRep representationUsingType:NSBitmapImageFileTypeJPEG properties:imageProps];
    return imageData;
}


- (NSImage *)generateThumbnail:(NSString *)imagePath toByte:(NSUInteger)maxLength {
    
    NSImage *image = [[NSImage alloc] initWithContentsOfFile:imagePath];
    NSData *imageData = [self getJPGImageData:image];
    
    // If the original image is already small, no thumbnail is needed
    if(imageData.length <= maxLength) {
        return nil;
    }
    
    NSImage *resultImage = image;
    NSUInteger lastDataLength = 0;
    
    while (imageData.length > maxLength && imageData.length != lastDataLength) {
        lastDataLength = imageData.length;
        CGFloat ratio = (CGFloat)maxLength / imageData.length;
        CGSize size = CGSizeMake((NSUInteger)(resultImage.size.width * sqrtf(ratio)),
                                 (NSUInteger)(resultImage.size.height * sqrtf(ratio)));// Use NSUInteger to prevent white blank
        
        NSImage *newImage = [[NSImage alloc] initWithSize:size];
        [newImage lockFocus];
        NSRect fromRect = NSMakeRect(0, 0, resultImage.size.width, resultImage.size.height);
        NSRect toRect = NSMakeRect(0, 0, size.width, size.height);
        [resultImage drawInRect:toRect fromRect:fromRect operation:NSCompositingOperationCopy fraction:1.0];
        [newImage unlockFocus];

        imageData = [self getJPGImageData:newImage];
        resultImage = newImage;
    }
    return resultImage;
}

- (void)sendRtmMessage:(AgoraRtmMessage *)rtmMessage {
    NSString *typeName = (self.mode.type == ChatTypePeer ? @"peer" : @"channel");
    
    __weak ChatViewController *weakSelf = self;
    void(^sent)(int) = ^(int status) {
        if (status != 0) {
            NSString *alert = [NSString stringWithFormat:@"send %@ message error: %d", typeName, status];
            [weakSelf showAlert:alert];
            return;
        }
        
        if(rtmMessage.type == AgoraRtmMessageTypeText) {
            [weakSelf appendMessage:rtmMessage.text mediaId:nil thumbnail:nil user:AgoraRtm.current];
            
        } else if(rtmMessage.type == AgoraRtmMessageTypeImage) {
            AgoraRtmImageMessage *imageMessage = (AgoraRtmImageMessage*)rtmMessage;
            [weakSelf appendMessage:nil mediaId:imageMessage.mediaId thumbnail:imageMessage.thumbnail user:AgoraRtm.current];
        }
    };
    
    switch (self.mode.type) {
        case ChatTypePeer: {
            AgoraRtmSendMessageOptions *option = [[AgoraRtmSendMessageOptions alloc] init];
            option.enableOfflineMessaging = [AgoraRtm oneToOneMessageType] == OneToOneMessageTypeOffline ? YES : NO;
            
            [AgoraRtm.kit sendMessage:rtmMessage toPeer:self.mode.name
                   sendMessageOptions:option
                           completion:^(AgoraRtmSendPeerMessageErrorCode errorCode) {
                
                sent((int)errorCode);
            }];
        }
        case ChatTypeGroup: {
            [self.rtmChannel sendMessage:rtmMessage completion:^(AgoraRtmSendChannelMessageErrorCode errorCode) {
                sent((int)errorCode);
            }];
        }
    }
}

#pragma mark - Channel
- (void)createChannel:(NSString *)channel {
    __weak ChatViewController *weakSelf = self;
    void(^errorHandle)(NSModalResponse) = ^(NSModalResponse returnCode) {
        [weakSelf.delegate chatVCWillClose:weakSelf];
    };
    
    AgoraRtmChannel *rtmChannel = [AgoraRtm.kit createChannelWithId:channel delegate:self];
    
    if (!rtmChannel) {
        [self showAlert:@"join channel fail" handle:errorHandle];
    }
    
    [rtmChannel joinWithCompletion:^(AgoraRtmJoinChannelErrorCode errorCode) {
        if (errorCode != AgoraRtmJoinChannelErrorOk) {
            NSString *alert = [NSString stringWithFormat:@"join channel error: %ld", errorCode];
            [weakSelf showAlert:alert handle:errorHandle];
        }
    }];
    
    self.rtmChannel = rtmChannel;
}

- (void)leaveChannel {
    if (self.mode.type == ChatTypePeer) {
        return;
    }
    
    [self.rtmChannel leaveWithCompletion:^(AgoraRtmLeaveChannelErrorCode errorCode) {
        NSLog(@"leave channel error: %ld", (long)errorCode);
    }];
}

#pragma mark - AgoraRtmDelegate
- (void)rtmKit:(AgoraRtmKit *)kit connectionStateChanged:(AgoraRtmConnectionState)state reason:(AgoraRtmConnectionChangeReason)reason {
    NSLog(@"connection state changed: %ld", (long)reason);
}

- (void)rtmKit:(AgoraRtmKit *)kit messageReceived:(AgoraRtmMessage *)message fromPeer:(NSString *)peerId {
    [self appendMessage:message.text mediaId:nil thumbnail:nil user:peerId];
}

- (void)rtmKit:(AgoraRtmKit *)kit imageMessageReceived:(AgoraRtmImageMessage *)message fromPeer:(NSString *)peerId {
    [self appendMessage:nil mediaId:message.mediaId thumbnail:message.thumbnail user:peerId];
}
#pragma mark - AgoraRtmChannelDelegate
- (void)channel:(AgoraRtmChannel *)channel memberJoined:(AgoraRtmMember *)member {
    NSString *user = member.userId;
    NSString *text = [user stringByAppendingString:@" join"];
    [self showAlert:text];
}

- (void)channel:(AgoraRtmChannel *)channel memberLeft:(AgoraRtmMember *)member {
    NSString *user = member.userId;
    NSString *text = [user stringByAppendingString:@" left"];
    [self showAlert:text];
}

- (void)channel:(AgoraRtmChannel *)channel messageReceived:(AgoraRtmMessage *)message fromMember:(AgoraRtmMember *)member {
    [self appendMessage:message.text mediaId:nil thumbnail:nil user:member.userId];
}

-(void)channel:(AgoraRtmChannel *)channel imageMessageReceived:(AgoraRtmImageMessage *)message fromMember:(AgoraRtmMember *)member {
    [self appendMessage:nil mediaId:message.mediaId thumbnail:message.thumbnail user:member.userId];
}

#pragma mark - UI
- (IBAction)doBackPressed:(NSButton *)sender {
    if (self.delegate && [self.delegate respondsToSelector:@selector(chatVCWillClose:)]) {
        [self.delegate chatVCWillClose:self];
    }
}

- (IBAction)doSendMsgPressed:(NSTextField *)sender {
    if ([self pressedReturnToSendText:sender.stringValue]) {
        sender.stringValue = @"";
    }
}

- (void)ifLoadOfflineMessages {
    switch (self.mode.type) {
        case ChatTypePeer: {
            NSArray *list = [AgoraRtm getOfflineMessagesFromUser:self.mode.name];
            
            for (AgoraRtmMessage *item in list) {
                if(item.type == AgoraRtmMessageTypeText){
                    [self appendMessage:item.text mediaId:nil thumbnail:nil user:self.mode.name];
                } else if(item.type == AgoraRtmMessageTypeImage){
                    AgoraRtmImageMessage *imageMessage = (AgoraRtmImageMessage*)item;
                    [self appendMessage:nil mediaId:imageMessage.mediaId thumbnail:imageMessage.thumbnail user:self.mode.name];
                }
            }
            
            [AgoraRtm removeOfflineMessageFromUser:self.mode.name];
        }
        default:
            break;
    }
}

- (void)appendMessage:(NSString *)text mediaId:(NSString*)mediaId thumbnail:(NSData*)thumbnail user:(NSString *)user {

    __weak ChatViewController *weakSelf = self;
    
    dispatch_async(dispatch_get_main_queue(), ^{
        Message *msg = [[Message alloc] init];
        msg.userId = user;
        msg.text = text;
        msg.mediaId = mediaId;
        msg.thumbnail = thumbnail;
        [weakSelf.list addObject:msg];
        
        if (weakSelf.list.count > 100) {
            [weakSelf.list removeObjectAtIndex:0];
        }
        
        NSInteger end = weakSelf.list.count - 1;
        [weakSelf.tableView reloadData];
        [weakSelf.tableView scrollRowToVisible:end];
    });
}

- (BOOL)pressedReturnToSendText:(NSString *)text {
    if (!text || text.length == 0) {
        return NO;
    }
    
    [self sendMessage:text];
    return YES;
}

#pragma mark - NSTableViewDataSource, NSTableViewDelegate
- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return self.list.count;
}

- (NSView *)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row {
    Message *msg = self.list[row];
    CellType type = [msg.userId isEqualToString:AgoraRtm.current] ? CellTypeRight : CellTypeLeft;
    if(msg.mediaId != nil) {
        ImageMessageCell *cell = [tableView makeViewWithIdentifier:@"ImageMessageCell" owner:nil];
        [cell updateType:type message:msg];
        return cell;
    } else {
        MessageCell *cell = [tableView makeViewWithIdentifier:@"MessageCell" owner:nil];
        [cell updateType:type message:msg];
        return cell;
    }
}

- (CGFloat)tableView:(NSTableView *)tableView heightOfRow:(NSInteger)row {
    Message *msg = self.list[row];
    if(msg.mediaId != nil) {
        return 212;
    }
    return 36;
}

- (void)tableViewSelectionDidChange:(NSNotification *)notification {
    NSInteger row = self.tableView.selectedRow;
    [self.tableView deselectRow:row];
}

- (NSMutableArray *)list {
    if (!_list) {
        _list = [NSMutableArray array];
    }
    return _list;
}

@end

File: Agora-RTM-Tutorial-macOS/Agora-RTM-Tutorial/ChatViewController.swift
//
//  ChatViewController.swift
//  Agora-Rtm-Tutorial
//
//  Created by CavanSu on 2019/1/17.
//  Copyright © 2019 Agora. All rights reserved.
//

import Cocoa
import AgoraRtmKit

enum ChatType {
    case peer(String), group(String)
    
    var description: String {
        switch self {
        case .peer:  return "peer"
        case .group: return "channel"
        }
    }
}

struct Message {
    var userId: String
    var text: String
}

protocol ChatVCDelegate: NSObjectProtocol {
    func chatVCWillClose(_ vc: ChatViewController)
    func chatVCBackToRootVC(_ vc: ChatViewController)
}

class ChatViewController: ShowAlertProtocol {
    @IBOutlet weak var inputTextField: NSTextField!
    @IBOutlet weak var tableView: NSTableView!
    
    lazy var list = [Message]()
    
    var delegate: ChatVCDelegate?
    var rtmChannel: AgoraRtmChannel?
    var type: ChatType = .peer("unknow")
    
    override func viewDidLoad() {
        ifLoadOfflineMessages()
        AgoraRtm.updateKit(delegate: self)
    }
    
    override func viewDidAppear() {
        switch type {
        case .peer(let name):     self.title = name
        case .group(let channel): self.title = channel; createChannel(channel)
        }
    }
    
    override func viewWillDisappear() {
        leaveChannel()
    }
    
    @IBAction func doSendMsgPressed(_ sender: NSTextField) {
        if pressedReturnToSendText(sender.stringValue) {
            sender.stringValue = ""
        }
    }
    
    @IBAction func doBackPressed(_ sender: NSButton) {
        self.delegate?.chatVCWillClose(self)
    }
}

// MARK: Send Message
private extension ChatViewController {
    func send(message: String, type: ChatType) {
        let sent = { [unowned self] (state: Int) in
            guard state == 0 else {
                self.showAlert("send \(type.description) message error: \(state)")
                return
            }
            
            guard let current = AgoraRtm.current else {
                return
            }
            self.appendMessage(user: current, content: message)
        }
        
        let rtmMessage = AgoraRtmMessage(text: message)
        
        switch type {
        case .peer(let name):
            let option = AgoraRtmSendMessageOptions()
            option.enableOfflineMessaging = (AgoraRtm.oneToOneMessageType == .offline ? true : false)
            
            AgoraRtm.kit?.send(rtmMessage, toPeer: name, sendMessageOptions: option, completion: { (error) in
                sent(error.rawValue)
            })
        case .group(_):
            rtmChannel?.send(rtmMessage) { (error) in
                sent(error.rawValue)
            }
        }
    }
}

// MARK: Chanel
private extension ChatViewController {
    func createChannel(_ channel: String) {
        let errorHandle = { [weak self] (response: NSApplication.ModalResponse) in
            guard let strongSelf = self else {
                return
            }
            strongSelf.delegate?.chatVCWillClose(strongSelf)
        }
        
        guard let rtmChannel = AgoraRtm.kit?.createChannel(withId: channel, delegate: self) else {
            showAlert("join channel fail", handler: errorHandle)
            return
        }
        
        rtmChannel.join { [weak self] (error) in
            if error != .channelErrorOk, let strongSelf = self {
                strongSelf.showAlert("join channel error: \(error.rawValue)", handler: errorHandle)
            }
        }
        
        self.rtmChannel = rtmChannel
    }
    
    func leaveChannel() {
        rtmChannel?.leave { (error) in
            print("leave channel error: \(error.rawValue)")
        }
    }
}

// MARK: AgoraRtmDelegate
extension ChatViewController: AgoraRtmDelegate {
    func rtmKit(_ kit: AgoraRtmKit, connectionStateChanged state: AgoraRtmConnectionState, reason: AgoraRtmConnectionChangeReason) {
        showAlert("connection state changed: \(state.rawValue)") { [weak self] (_) in
            if reason == .remoteLogin, let strongSelf = self {
                strongSelf.delegate?.chatVCBackToRootVC(strongSelf)
            }
        }
    }
    
    func rtmKit(_ kit: AgoraRtmKit, messageReceived message: AgoraRtmMessage, fromPeer peerId: String) {
        appendMessage(user: peerId, content: message.text)
    }
}

// MARK: AgoraRtmChannelDelegate
extension ChatViewController: AgoraRtmChannelDelegate {
    func channel(_ channel: AgoraRtmChannel, memberJoined member: AgoraRtmMember) {
        DispatchQueue.main.async { [unowned self] in
            self.showAlert("\(member.userId) join")
        }
    }
    
    func channel(_ channel: AgoraRtmChannel, memberLeft member: AgoraRtmMember) {
        DispatchQueue.main.async { [unowned self] in
            self.showAlert("\(member.userId) left")
        }
    }
    
    func channel(_ channel: AgoraRtmChannel, messageReceived message: AgoraRtmMessage, from member: AgoraRtmMember) {
        appendMessage(user: member.userId, content: message.text)
    }
}

private extension ChatViewController {
    func ifLoadOfflineMessages() {
        switch type {
        case .peer(let name):
            guard let messages = AgoraRtm.getOfflineMessages(from: name) else {
                return
            }
            
            for item in messages {
                appendMessage(user: name, content: item.text)
            }
            
            AgoraRtm.removeOfflineMessages(from: name)
        default:
            break
        }
    }
    
    func pressedReturnToSendText(_ text: String?) -> Bool {
        guard let text = text, text.count > 0 else {
            return false
        }
        send(message: text, type: type)
        return true
    }
    
    func appendMessage(user: String, content: String) {
        DispatchQueue.main.async { [unowned self] in
            let msg = Message(userId: user, text: content)
            self.list.append(msg)
            if self.list.count > 100 {
                self.list.removeFirst()
            }
            let end = self.list.count - 1
            
            self.tableView.reloadData()
            self.tableView.scrollRowToVisible(end)
        }
    }
}

extension ChatViewController: NSTableViewDelegate, NSTableViewDataSource {
    func numberOfRows(in tableView: NSTableView) -> Int {
        return list.count
    }
    
    func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
        let msg = list[row]
        let type: CellType = msg.userId == AgoraRtm.current ? .right : .left
        let cell = tableView.makeView(withIdentifier: NSUserInterfaceItemIdentifier(rawValue: "MessageCell"), owner: nil) as! MessageCell
        cell.update(type: type, message: msg)
        return cell
    }
    
    func tableView(_ tableView: NSTableView, heightOfRow row: Int) -> CGFloat {
        return 36
    }
}
